Packages/com.unity.inputsystem/InputSystem/Plugins/iOS/iOSScreenKeyboard.mm

```diff
@@ -0,0 +1,657 @@
+        [[NSNotificationCenter defaultCenter] addObserver: self selector: @selector(keyboardDidShow:) name: UIKeyboardDidShowNotification object: nil];
+        [[NSNotificationCenter defaultCenter] addObserver: self selector: @selector(keyboardWillHide:) name: UIKeyboardWillHideNotification object: nil];
+        [[NSNotificationCenter defaultCenter] addObserver: self selector: @selector(keyboardDidChangeFrame:) name: UIKeyboardDidChangeFrameNotification object: nil];
+#endif
+
+        [[NSNotificationCenter defaultCenter] addObserver: self selector: @selector(textInputDone:) name: UITextFieldTextDidEndEditingNotification object: nil];
+    }
+
+    return self;
+}
+
+- (void)dealloc
+{
+    KEYBOARD_LOG(@"dealloc");
+    if (m_State == StateVisible)
+    {
+        m_State = StateDone;
+        [self hideUIDelayedWithCallback];
+    }
+#if PLATFORM_IOS
+    [[NSNotificationCenter defaultCenter] removeObserver: self name: UIKeyboardWillShowNotification object: nil];
+    [[NSNotificationCenter defaultCenter] removeObserver: self name: UIKeyboardDidShowNotification object: nil];
+    [[NSNotificationCenter defaultCenter] removeObserver: self name: UIKeyboardWillHideNotification object: nil];
+    [[NSNotificationCenter defaultCenter] removeObserver: self name: UIKeyboardDidChangeFrameNotification object: nil];
+#endif
+    if ([iOSScreenKeyboardBridge simulateTextSelection])
+    {
+        [m_TextField removeObserver: self forKeyPath: @"selectedTextRange"];
+    }
+
+    [[NSNotificationCenter defaultCenter] removeObserver: self name: UITextFieldTextDidEndEditingNotification object: nil];
+}
+
+- (void)setTextInputTraits:(id<UITextInputTraits>)traits
+    withParam:(iOSScreenKeyboardShowParamsNative)param
+    withCap:(UITextAutocapitalizationType)capitalization
+{
+    traits.keyboardType = param.keyboardType;
+    traits.autocorrectionType = param.autocorrectionType;
+    traits.secureTextEntry = param.secure;
+    traits.keyboardAppearance = param.appearance;
+    traits.autocapitalizationType = capitalization;
+}
+
+- (void)hideUIDelayed
+{
+    KEYBOARD_LOG(@"hideUIDelayed");
+    [m_InputView resignFirstResponder];
+
+    [m_EditView removeFromSuperview];
+    m_EditView.hidden = YES;
+}
+
+- (void)hideUIDelayedWithCallback
+{
+    KEYBOARD_LOG(@"hideUIDelayedWithCallback");
+    [self hideUIDelayed];
+    m_ShowParams.callbacks.stateChangedCallback(m_State);
+}
+
+- (void)systemHideKeyboard
+{
+    KEYBOARD_LOG(@"systemHideKeyboard");
+
+    m_Active = m_EditView.isFirstResponder;
+    m_EditView.hidden = YES;
+
+    m_Area = CGRectMake(0, 0, 0, 0);
+}
+
+#if PLATFORM_IOS
+- (void)positionInput:(CGRect)kbRect x:(float)x y:(float)y
+{
+    float safeAreaInsetLeft = 0;
+    float safeAreaInsetRight = 0;
+
+#if UNITY_HAS_IOSSDK_11_0
+    if (@available(iOS 11.0, *))
+    {
+        safeAreaInsetLeft = [UnityGetGLView() safeAreaInsets].left;
```

Comment:
Correct me if I'm wrong, but `UnityGetGLView` seems to be an internal unity api. So someone like me can land a change into trunk changing this api's, which at best will results in this code not compiling, at worst working incorrectly. To make a backend for native keyboard robust and exist outside of trunk, IMHO we should make it not use any of internal api's. And we should have some way to test that it works on CI.